_posts/2016-11-09-Rust-1.13.md

```diff
@@ -0,0 +1,469 @@
+anyway. While we're [still working on improving the
+diagnostic](https://github.com/rust-lang/rust/issues/35946), we now have
+the ability to do so.
+
+#### Performance improvements
+
+
+There has bee a lot of focus on compiler performance lately. There's good news
+in this release, and more to come.
+
+Mark Simulacrum and Nick Cameron have been refining [perf.rust-lang.org], our
+tool for tracking compiler performance. It runs the [rustc-benchmarks] suite
+regularly, on dedicated hardware, and tracks the results over time. This tool
+records the results for each pass in the compiler and is used by the compiler
+developers to narrow commit ranges of performance regressions. It's an important
+part of our toolbox!
+
+We can use this tool to look at a [graph] of performance over the 1.13
+development cycle. This cycle covered the dates from August 16 through September
+29 (the graph begins from Augest 25th though and is filtered in a few ways to
+eliminate bogus, incomplete, or confusing results). There appear to be some big
+reductions, which are quantified on the corresponding [statistics] page.
+
+The big improvement demonstrated in the graphs, on September 1, is from an
+optimization from Niko to [cache normalized projections during
+translation][cache]. That is to say, during generation of LLVM IR, the compiler
+no longer recomputes concrete instances of associated types each time they are
+needed, but instead reuses previously-computed values. This optimization doesn't
+affect all code bases, but in code bases that exhibit certain patterns, like
+[futures-rs], where [debug mode build-time improved by up to 40%][ev1], you'll notice
+the difference.
+
+Another such optimization, that doesn't affect every crate but does affect some
+in a big way, came from Michael Woerister, and improves compile time for crates
+that export many [inline] functions. When a function is marked `#[inline]`, in
+addition to translating that function for use by the current crate, the compiler
+stores its MIR representation in the crate rlib, and translates the function to
+LLVM IR in every crate that calls it. The optimization Michael did is obvious in
+retrospect: there are some cases where inline functions are only for the
+consumption of other crates, and never called from the crate in which they are
+defined; so the compiler doesn't need to translate code for inline functions in
+the crate they are defined _unless_ they are called directly. This saves the
+cost of rustc converting the function to LLVM IR and LLVM optimizing and
+converting the function to machine code (TODO is this right about converting to
+machine code - were these functions public before?).
+
+In some cases this results in dramatic improvements. Build times for the ndarray
+crate [improved by 50%][ev2], and in the (unreleased) [winapi 0.3] crate, which
+is nothing but inline functions, rustc now generates the following LLVM IR:
```

Comment:
To clarify, it has plenty of things other than inline functions, its just that all the functions are inline.